components/filter/filter.js

```diff
@@ -0,0 +1,253 @@
+		return [bodyStandardStyles, css`
+			div[slot="header"] {
+				padding: 18px 6px;
+			}
+			.d2l-filter-dimension-header {
+				align-items: center;
+				display: flex;
+				width: 100%;
+			}
+			.d2l-filter-dimension-header-text {
+				padding-right: calc(2rem + 2px);
+				text-align: center;
+				width: 100%;
+			}
+			:host([dir="rtl"]) .d2l-filter-dimension-header-text {
+				padding-left: calc(2rem + 2px);
+				padding-right: 0;
+			}
+
+			.d2l-filter-dimension-set-value-text {
+				overflow: hidden;
+				text-overflow: ellipsis;
+				white-space: nowrap;
+			}
+		`];
+	}
+
+	constructor() {
+		super();
+		this.disabled = false;
+		this._dimensions = [];
+	}
+
+	firstUpdated(changedProperties) {
+		super.firstUpdated(changedProperties);
+		this.addEventListener('d2l-filter-dimension-data-change', this._handleDimensionDataChange);
+
+		// Prevent these events from bubbling out of the filter
+		this.addEventListener('d2l-hierarchical-view-hide-complete', this._stopPropagation);
+		this.addEventListener('d2l-hierarchical-view-hide-start', this._stopPropagation);
+		this.addEventListener('d2l-hierarchical-view-show-complete', this._stopPropagation);
+		this.addEventListener('d2l-hierarchical-view-show-start', this._stopPropagation);
+		this.addEventListener('d2l-hierarchical-view-resize', this._stopPropagation);
+	}
+
+	render() {
+		const header = this._buildHeader();
+		const dimensions = this._buildDimensions();
+
+		return html`
+			<d2l-dropdown-button-subtle
+				@d2l-dropdown-close="${this._handleDropdownClose}"
+				@d2l-dropdown-open="${this._stopPropagation}"
+				@d2l-dropdown-position="${this._stopPropagation}"
+				text="${this.localize('components.filter.filters')}"
+				?disabled="${this.disabled}">
+				<d2l-dropdown-menu min-width="285" max-width="420" no-padding-header>
+					${header}
+					<d2l-menu label="${this.localize('components.filter.filters')}">
+						${dimensions}
+					</d2l-menu>
+				</d2l-dropdown-menu>
+			</d2l-dropdown-button-subtle>
+			<slot @slotchange="${this._handleSlotChange}"></slot>
+		`;
+	}
+
+	_buildDimension(dimension) {
+		let dimensionHTML;
+		switch (dimension.type) {
+			case 'd2l-filter-dimension-set':
+				dimensionHTML = this._createSetDimension(dimension);
+				break;
+		}
+		return html`
+			<d2l-hierarchical-view
+				@d2l-hierarchical-view-show-complete="${this._handleDimensionShowComplete}"
+				@d2l-hierarchical-view-show-start="${this._handleDimensionShowStart}"
+				data-key="${dimension.key}">
+				${dimensionHTML}
+			</d2l-hierarchical-view>
+		`;
+	}
+
+	_buildDimensions() {
+		return this._dimensions.map((dimension) => {
+			const builtDimension = this._buildDimension(dimension);
+			return html`<d2l-menu-item text="${dimension.text}">
+				${builtDimension}
+			</d2l-menu-item>`;
+		});
+	}
+
+	_buildHeader() {
+		if (!this._activeDimensionKey) return null;
+
+		const dimensionText = this._dimensions.find(dimension => dimension.key === this._activeDimensionKey).text;
```

Comment:
Not sure if it's possible for the dimension to not be found... maybe they dynamically removed a dimension? 🤷

Reply:
Yeah like there _should_ always be a dimension here, because if we're rendering a header they _just_ clicked that dimension.  But then maybe it got dynamically removed after that?  There were a couple places where I ran into this conceptual problem - there _should_ be a dimension, so is it better to let it error out and show up in our js logs so we can fix the place that's adding/removing dimensions weirdly?  Or should we err on the side of trying to gracefully handle those weird scenarios?

Reply:
Yeah seems like not worth worrying about unless you notice it as an issue in practice. 👍